File: modules/device-attitude/android/src/main/java/expo/modules/deviceattitude/DeviceAttitudeModule.kt
package expo.modules.deviceattitude

import android.content.Context
import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.hardware.SensorManager
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import kotlin.math.asin
import kotlin.math.atan2

private const val EVENT_ATTITUDE_DID_UPDATE = "onAttitudeUpdate"
private const val DEFAULT_UPDATE_INTERVAL_US = 33_000 // ~30Hz

// Koristi TYPE_ROTATION_VECTOR (zirokop+akcelerometar+kompas, fuzionisano na
// nivou OS-a/senzorskog hardvera) umjesto rucnog EMA zaglađivanja nad
// odvojenim kompas/DeviceMotion ocitanjima -- ovo je isti pristup koji
// koriste prave AR aplikacije za nebo (npr. Star Walk).
class DeviceAttitudeModule : Module() {
  private val sensorManager: SensorManager?
    get() = appContext.reactContext?.getSystemService(Context.SENSOR_SERVICE) as? SensorManager

  private val rotationMatrix = FloatArray(9)
  private var intervalUs = DEFAULT_UPDATE_INTERVAL_US

  private val listener = object : SensorEventListener {
    override fun onSensorChanged(event: SensorEvent) {
      SensorManager.getRotationMatrixFromVector(rotationMatrix, event.values)

      // Android-ov getRotationMatrixFromVector vec direktno preslikava
      // device->world (world = R * device, world frame je X=istok, Y=sjever,
      // Z=gore) -- za razliku od iOS-a, ovdje NIJE potrebna transpozicija.
      // Kamera (zadnja strana telefona) gleda u device -Z pravcu.
      val east = -rotationMatrix[2]
      val north = -rotationMatrix[5]
      val up = (-rotationMatrix[8]).coerceIn(-1f, 1f)

      val altitude = Math.toDegrees(asin(up.toDouble()))
      var azimuth = Math.toDegrees(atan2(east.toDouble(), north.toDouble()))
      if (azimuth < 0) azimuth += 360

      sendEvent(EVENT_ATTITUDE_DID_UPDATE, mapOf("azimuth" to azimuth, "altitude" to altitude))
    }

    override fun onAccuracyChanged(sensor: Sensor?, accuracy: Int) = Unit
  }

  override fun definition() = ModuleDefinition {
    Name("DeviceAttitude")

    Events(EVENT_ATTITUDE_DID_UPDATE)

    AsyncFunction("isAvailableAsync") {
      sensorManager?.getDefaultSensor(Sensor.TYPE_ROTATION_VECTOR) != null
    }

    AsyncFunction("setUpdateInterval") { intervalMs: Double ->
      intervalUs = (intervalMs * 1000).toInt()
    }

    OnStartObserving {
      val sensor = sensorManager?.getDefaultSensor(Sensor.TYPE_ROTATION_VECTOR) ?: return@OnStartObserving
      sensorManager?.registerListener(listener, sensor, intervalUs)
    }

    OnStopObserving {
      sensorManager?.unregisterListener(listener)
    }

    OnDestroy {
      sensorManager?.unregisterListener(listener)
    }
  }
}

File: modules/device-attitude/ios/DeviceAttitudeModule.swift
import CoreMotion
import ExpoModulesCore

private let EVENT_ATTITUDE_DID_UPDATE = "onAttitudeUpdate"
private let DEFAULT_UPDATE_INTERVAL_SECONDS = 1.0 / 30.0

// Koristi CMDeviceMotion (zirokop+akcelerometar+kompas, fuzionisano na nivou
// OS-a preko CMAttitudeReferenceFrameXTrueNorthZVertical) umjesto rucnog
// EMA zaglađivanja nad odvojenim kompas/DeviceMotion ocitanjima -- ovo je isti
// pristup koji koriste prave AR aplikacije za nebo (npr. Star Walk).
public final class DeviceAttitudeModule: Module {
  private lazy var motionManager = CMMotionManager()
  private lazy var operationQueue = OperationQueue()

  public func definition() -> ModuleDefinition {
    Name("DeviceAttitude")

    Events(EVENT_ATTITUDE_DID_UPDATE)

    AsyncFunction("isAvailableAsync") {
      motionManager.isDeviceMotionAvailable
    }

    AsyncFunction("setUpdateInterval") { (intervalMs: Double) in
      motionManager.deviceMotionUpdateInterval = intervalMs / 1000.0
    }

    OnStartObserving {
      startUpdates()
    }

    OnStopObserving {
      motionManager.stopDeviceMotionUpdates()
    }

    OnDestroy {
      motionManager.stopDeviceMotionUpdates()
    }
  }

  private func startUpdates() {
    guard motionManager.isDeviceMotionAvailable, !motionManager.isDeviceMotionActive else {
      return
    }
    if motionManager.deviceMotionUpdateInterval <= 0 {
      motionManager.deviceMotionUpdateInterval = DEFAULT_UPDATE_INTERVAL_SECONDS
    }
    motionManager.startDeviceMotionUpdates(using: .xTrueNorthZVertical, to: operationQueue) { [weak self] data, _ in
      guard let self, let data else { return }
      let r = data.attitude.rotationMatrix

      // VAZNO: CMAttitude.rotationMatrix je transponovana verzija matrice koja
      // preslikava device->world (Apple je vraca kao world->device -- potvrdjeno
      // iz vise nezavisnih izvora, vidi PR opis). Zato ovdje koristimo TRECI RED
      // (m31, m32, m33), ne trecu kolonu, da dobijemo pravac zadnje kamere
      // (device -Z osa, "iza ekrana") u world koordinatama, gdje je za
      // .xTrueNorthZVertical referentni okvir: X=sjever, Y=zapad, Z=gore.
      // Ako se na uredjaju pokaze da je sjever/jug ili istok/zapad zamijenjen,
      // popravka je zamjena predznaka/redova ovdje, ne u JS sloju.
      let north = -r.m31
      let east = r.m32
      let up = max(-1, min(1, -r.m33))

      let altitude = asin(up) * 180 / .pi
      var azimuth = atan2(east, north) * 180 / .pi
      if azimuth < 0 { azimuth += 360 }

      self.sendEvent(EVENT_ATTITUDE_DID_UPDATE, [
        "azimuth": azimuth,
        "altitude": altitude
      ])
    }
  }
}
